### ISSUE TITLE
Enumerable.Last[TSource] (System.Collections.Generic.IEnumerable`1[T] source)

### ISSUE THREAD

**Issue by PaulsonMac** [xamarin/Xamarin.Forms#6700]
### Description
System.InvalidOperationException: Sequence contains no elements

### Platform
- Android (Versions 8, 8.1, 9)

### Reference Link

#5857, #4715 

### Issue First reported 
12 Dec 2018

Kindly fix this. This crash occurs very frequently. (CRITICAL ISSUE)

**Comment by samhouts** [comment]
@PaulsonMac  Can you please attach a small project that demonstrates this issue? Thanks!

**Comment by samhouts** [comment]
@PaulsonMac  Since we haven't heard from you in more than 30 days, we hope this issue is no longer affecting you. If it is, please reopen this issue and provide the requested information so that we can look into it further. Thank you!

**Comment by GiridharanNarayanan** [comment]
@samhouts , I am seeing this issue too. Can I share a project and will you be able to get some help on this?

**Comment by samhouts** [comment]
@GiridharanNarayanan Yes, please do! Thank you.

**Comment by acaliaro** [comment]
I think to have the same problem but I am not able to reproduce it

Package: com.XXX.XXXXXX
Version Code: 64
Version Name: 2.0.0
Android: 6.0.1
Android Build: 01-21-04.1-MN-00
Manufacturer: Zebra Technologies
Model: TC51
CrashReporter Key: 7c562aad-e1a9-47e6-8040-d9a82a160020
Date: undefined

Xamarin Exception Stack:
System.InvalidOperationException: Sequence contains no elements
  at System.Linq.Enumerable.Last[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00010] in <6031b424a9974301bc948da4dd4015d8>:0
  at Xamarin.Forms.Platform.Android.AppCompat.Platform.HandleBackPressed (System.Object sender, System.EventArgs e) [0x00015] in <b356e1b6c53b471db4568de2eac3880b>:0
  at Xamarin.Forms.Platform.Android.FormsAppCompatActivity.OnBackPressed () [0x00007] in <b356e1b6c53b471db4568de2eac3880b>:0
  at Android.App.Activity.n_OnBackPressed (System.IntPtr jnienv, System.IntPtr native__this) [0x00009] in <63f6c3d513ec4ea5bd8cf6aba6672b75>:0
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.12(intptr,intptr)

Thread 1:
0   dalvik.system.VMStack.getThreadStackTrace(VMStack.java:-2)
1   java.lang.Thread.getStackTrace(Thread.java:580)
2   java.lang.Thread.getAllStackTraces(Thread.java:522)
3   com.microsoft.appcenter.crashes.Crashes.saveUncaughtException(Crashes.java:970)
4   com.microsoft.appcenter.crashes.WrapperSdkExceptionManager.saveWrapperException(WrapperSdkExceptionManager.java:55)
5   md51558244f76c53b6aeda52c8a337f2c37.FormsAppCompatActivity.n_onBackPressed(FormsAppCompatActivity.java:-2)
6   md51558244f76c53b6aeda52c8a337f2c37.FormsAppCompatActivity.onBackPressed(FormsAppCompatActivity.java:40)
7   android.app.Activity.onKeyUp(Activity.java:2495)
8   android.view.KeyEvent.dispatch(KeyEvent.java:2766)
9   android.support.v4.view.KeyEventDispatcher.activitySuperDispatchKeyEventPre28(KeyEventDispatcher.java:137)
10  android.support.v4.view.KeyEventDispatcher.dispatchKeyEvent(KeyEventDispatcher.java:87)
11  android.support.v4.app.SupportActivity.dispatchKeyEvent(ComponentActivity.java:126)
12  android.support.v7.app.AppCompatActivity.dispatchKeyEvent(AppCompatActivity.java:535)
13  android.support.v7.view.WindowCallbackWrapper.dispatchKeyEvent(WindowCallbackWrapper.java:59)
14  android.support.v7.app.AppCompatDelegateImpl$AppCompatWindowCallback.dispatchKeyEvent(AppCompatDelegateImpl.java:2533)
15  android.support.v7.view.WindowCallbackWrapper.dispatchKeyEvent(WindowCallbackWrapper.java:59)
16  com.android.internal.policy.PhoneWindow$DecorView.dispatchKeyEvent(PhoneWindow.java:2310)
17  android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:4127)
18  android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4089)
19  android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3642)
20  android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3695)
21  android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3661)
22  android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:3787)
23  android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3669)
24  android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:3844)
25  android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3642)
26  android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3695)
27  android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3661)
28  android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3669)
29  android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3642)
30  android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3695)
31  android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3661)
32  android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:3820)
33  android.view.ViewRootImpl$ImeInputStage.onFinishedInputEvent(ViewRootImpl.java:3981)
34  android.view.inputmethod.InputMethodManager$PendingEvent.run(InputMethodManager.java:2253)
35  android.view.inputmethod.InputMethodManager.invokeFinishedInputEventCallback(InputMethodManager.java:1874)
36  android.view.inputmethod.InputMethodManager.finishedInputEvent(InputMethodManager.java:1865)
37  android.view.inputmethod.InputMethodManager$ImeInputEventSender.onInputEventFinished(InputMethodManager.java:2230)
38  android.view.InputEventSender.dispatchInputEventFinished(InputEventSender.java:141)
39  android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
40  android.os.MessageQueue.next(MessageQueue.java:323)
41  android.os.Looper.loop(Looper.java:135)
42  android.app.ActivityThread.main(ActivityThread.java:5417)
43  java.lang.reflect.Method.invoke(Method.java:-2)
44  com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
45  com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

Thread 500:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Daemons$ReferenceQueueDaemon.run(Daemons.java:147)
2   java.lang.Thread.run(Thread.java:818)

Thread 501:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Object.wait(Object.java:423)
2   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:101)
3   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:72)
4   java.lang.Daemons$FinalizerDaemon.run(Daemons.java:185)
5   java.lang.Thread.run(Thread.java:818)

Thread 502:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Daemons$FinalizerWatchdogDaemon.waitForObject(Daemons.java:255)
2   java.lang.Daemons$FinalizerWatchdogDaemon.run(Daemons.java:227)
3   java.lang.Thread.run(Thread.java:818)

Thread 503:
0   dalvik.system.VMRuntime.runHeapTasks(VMRuntime.java:-2)
1   java.lang.Daemons$HeapTaskDaemon.run(Daemons.java:355)
2   java.lang.Thread.run(Thread.java:818)

Thread 510:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:323)
2   android.os.Looper.loop(Looper.java:135)
3   android.os.HandlerThread.run(HandlerThread.java:61)

Thread 513:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Thread.parkFor$(Thread.java:1220)
2   sun.misc.Unsafe.park(Unsafe.java:299)
3   java.util.concurrent.locks.LockSupport.park(LockSupport.java:158)
4   java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2013)
5   java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:410)
6   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1036)
7   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1098)
8   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
9   java.lang.Thread.run(Thread.java:818)

Thread 515:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:323)
2   android.os.Looper.loop(Looper.java:135)
3   android.os.HandlerThread.run(HandlerThread.java:61)

Thread 518:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Thread.parkFor$(Thread.java:1220)
2   sun.misc.Unsafe.park(Unsafe.java:299)
3   java.util.concurrent.locks.LockSupport.park(LockSupport.java:158)
4   java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2013)
5   java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:410)
6   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1036)
7   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1098)
8   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
9   java.lang.Thread.run(Thread.java:818)

Thread 519:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Thread.parkFor$(Thread.java:1220)
2   sun.misc.Unsafe.park(Unsafe.java:299)
3   java.util.concurrent.locks.LockSupport.park(LockSupport.java:158)
4   java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2013)
5   java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:410)
6   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1036)
7   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1098)
8   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
9   java.lang.Thread.run(Thread.java:818)

Thread 520:
0   android.net.LocalSocketImpl.readba_native(LocalSocketImpl.java:-2)
1   android.net.LocalSocketImpl.access$400(LocalSocketImpl.java:33)
2   android.net.LocalSocketImpl$SocketInputStream.read(LocalSocketImpl.java:101)
3   com.qti.tcmclient.DpmTcmClient$TcmReceiver.run(DpmTcmClient.java:140)
4   java.lang.Thread.run(Thread.java:818)

Thread 521:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.okio.AsyncTimeout.awaitTimeout(AsyncTimeout.java:311)
2   com.android.okhttp.okio.AsyncTimeout.access$000(AsyncTimeout.java:40)
3   com.android.okhttp.okio.AsyncTimeout$Watchdog.run(AsyncTimeout.java:286)

Thread 522:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.ConnectionPool.performCleanup(ConnectionPool.java:307)
2   com.android.okhttp.ConnectionPool.runCleanupUntilPoolIsEmpty(ConnectionPool.java:244)
3   com.android.okhttp.ConnectionPool.access$000(ConnectionPool.java:54)
4   com.android.okhttp.ConnectionPool$1.run(ConnectionPool.java:98)
5   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
6   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
7   java.lang.Thread.run(Thread.java:818)

**Comment by 15mgm15** [comment]
@acaliaro By any chance do you have a solution for this?

**Comment by acaliaro** [comment]
> @acaliaro By any chance do you have a solution for this?

absolutely no

**Comment by samhouts** [comment]
Can you please attach a small project that demonstrates this issue? Thanks!

**Comment by acaliaro** [comment]
I can’t 
Mine is a very big project and it’s in production all over the world in different Outlets. I really don’t know how to reproduce it

**Comment by 15mgm15** [comment]
I was able to find the issue and it was caused by having two NavigationPages:

```
MainPage = new NavigationPage(new TabbedPage)
class TabbedPage { Children = { new NavigationPage { new ContentPage }, new NavigationPage { new ContentPage }, new NavigationPage { new ContentPage } } }
```

**Comment by acaliaro** [comment]
> I was able to find the issue and it was caused by having two NavigationPages:
> MainPage = new NavigationPage(new TabbedPage)
> class TabbedPage { Children = { new NavigationPage { new ContentPage }, new NavigationPage { new ContentPage }, new NavigationPage { new ContentPage } } }

why do you say that it's related to this?

**Comment by 15mgm15** [comment]
@acaliaro Actually I don't know now, I thought it was related to having a `TabbedPage` with `NavigationPages` in it but now that I am using the or at least I think I am using the correct navigation context I am still experiencing this issue with the first child of the `TabbedPage` anyways... still investigating.

**Comment by 15mgm15** [comment]
For the next project, I will definitely consider using an MVVM Framework.

**Comment by acaliaro** [comment]
> For the next project, I will definitely consider using an MVVM Framework.

Yes you should, you could use Xamarin integrated MVVM

**Comment by 15mgm15** [comment]
@acaliaro I was thinking more like PRISM, or maybe just only a IoC container like autofac

**Comment by dgandhi17** [comment]
I am able to trace this issue in Xamarin.Forms source code, and found out that below method in "Platform.cs" in namespace "Xamarin.Forms.Platform.Android.AppCompat"

```
bool HandleBackPressed(object sender, EventArgs e)
		{
			if (NavAnimationInProgress)
				return true;

			Page root = _navModel.Roots.Last();
			bool handled = root.SendBackButtonPressed();

			return handled;
		}
```
"Page root = _navModel.Roots.Last();" Here _navMOdel.Roots is empty and when it say that return Last() it will throw "System.InvalidOperationException: Sequence contains no elements"

we may need to change this implementation to return false if Roots is empty

**Comment by dgandhi17** [comment]
@samhouts Can you verify above comment? as I am not able to build xamarin.forms solution so can't create PR for the same.

**Comment by jfversluis** [comment]
@dgandhi17 you say you were able to trace this issue. Does that mean you have some way of reproducing it?

Although the fix might seem easy to just handle the `Last()` call there in a better way, it does mean that the app is in some weird navigation state. So even if we might fix it that way, there is no way to tell what the behavior will be with that new code in place. It might not actually crash anymore, but it might cause the end-user to not see any responses to navigation events anymore for example.

To see how a possible fix might behave it is very valuable to have a reproduction in this case.

**Comment by dgandhi17** [comment]
@jfversluis Well if we say that the method's responsibility is to send backpress event to the last page if it is available and return true because we handled it. But it should not trigger the back press event for the page which is not there in the navigation stack and should return false.

So here my suggestion is that in FormsAppCompactActivity.cs OnBackPressed() method it should be handled like if we don't have any page we should exit from the application as mostly forms apps are like single activity app, like for default Android native apps for single activity pressing back button exit the user from App.

**Comment by jfversluis** [comment]
I totally agree, and I will open a PR to add some defensive code around this so we can discuss it also with other team members, but that was not my question :)

Have you found some way to reproduce it consistently? That way we won't have to guess what could or could not happen, but we can test it and also write a test to make sure it never happens again.

**Comment by samhouts** [comment]
closed by #8666
